pool.go

```diff
@@ -208,50 +229,100 @@ func (p *pool) fetchWith(ctx context.Context, c cid.Cid, with string) (blk block
 	for i := 0; i < len(nodes); i++ {
 		blk, err = p.doFetch(ctx, nodes[i], c)
 		if err != nil {
-			p.lk.RLock()
-			idx := -1
-			var nm *Member
-			var needUpdate bool
-			for j, m := range p.endpoints {
-				if m.String() == nodes[i] {
-					if nm, needUpdate = m.Downvote(p.config.PoolFailureDownvoteDebounce); needUpdate {
-						idx = j
+			goLogger.Debugw("fetch failed", "from", nodes[i], "of", c, "attempt", i, "error", err)
+		}
+		var idx int
+		var nm *Member
+
+		if err == nil {
+			// we need to acquire the lock to update the member's weight in the pool.
+			// and now that we are here, we know we're gonna return and hit the defer before exiting this block.
+			p.lk.Lock()
+			defer p.lk.Unlock()
+
+			// Saturn fetch worked, we should try upvoting the member.
+			idx, nm = p.updateWeightUnlocked(nodes[i], func(old int) int {
+				// bump by 20 percent only if the current replication factor is less than 20.
+				if old < defaultReplication {
+					updated := old + (old / 5)
+					if updated > defaultReplication {
+						updated = defaultReplication
 					}
-					break
+					return updated
 				}
+				return old
+			})
+
+			if idx != -1 && nm != nil && p.endpoints[idx].url == nm.url {
+				p.endpoints[idx] = nm
+				p.c.UpdateWithWeights(p.endpoints.ToWeights())
 			}
-			p.lk.RUnlock()
-			if idx != -1 {
-				p.lk.Lock()
-				if p.endpoints[idx].string == nm.string {
-					if nm.replication == 0 {
-						p.c = p.c.RemoveNode(nm.string)
-						p.endpoints = append(p.endpoints[:idx], p.endpoints[idx+1:]...)
-						if len(p.endpoints) < p.config.PoolLowWatermark {
-							select {
-							case p.refresh <- struct{}{}:
-							default:
-							}
-						}
-					} else {
-						p.endpoints[idx] = nm
-						p.c.UpdateWithWeights(p.endpoints.ToWeights())
+
+			// Saturn fetch worked, we return the block.
+			return
+		}
+
+		p.lk.RLock()
+		// Saturn fetch failed, we downvote the failing member and try the next one.
+		idx, nm = p.updateWeightUnlocked(nodes[i], func(old int) int {
+			return old / 2
+		})
+		p.lk.RUnlock()
+
+		// we weren't able to downvote the failing Saturn node, let's just retry the fetch with a new node.
+		if idx == -1 || nm == nil {
+			continue
+		}
+
+		// we need to take the lock as we're updating the list of pool endpoint members below.
+		p.lk.Lock()
+		if p.endpoints[idx].url == nm.url {
```

Comment:
:warning:  @aarshkshah1992 mind checking if my understanding is correct that this line is causing random panics? 
Details in  https://github.com/filecoin-saturn/caboose/issues/29

Reply:
i think this is because between the `Runlock` and the `Lock` the `endpoints` can change to a new set. we should find `idx` within the same critical section for stability

Reply:
#30 should fix